llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -21553,6 +21601,15 @@ bool DAGCombiner::tryStoreMergeOfLoads(SmallVectorImpl<MemOpLink> &StoreNodes,
       JointMemOpVT = EVT::getIntegerVT(Context, SizeInBits);
     }
 
+    // Check if there is a call in the load/store chain.
```

Comment:
I just noticed there is a better integration here.  The way you have this phrased, we either can merge to the vector type, or we don't merge at all.

I think what you actually want to do here is to add a couple parameters to the existing canMergeStoresTo interface.  If you add both the SrcVT, and a "IsOverCall" boolean, then that interface returning the answer suggested above in the IsOverCall case will result in a wider scalar type being chosen if one is available.

Reply:
Though, I'll note I'm also fine in this being a follow up change.  I'm happy to pick that up if you want to focus on just getting this in.

Reply:
I'm happy with either option: to land this PR and do a follow-up patch, or work on this one. Up to you.

Also, I did a couple of tests with your first suggestion on checking for CSR for a given type in https://github.com/mikhailramalho/llvm-project/commit/df27cb2a90aeb62cce46ddac28fbee88b698926f, should I give up on that approach? There are still a couple of crashes I need to debug.

Reply:
Let's go for minimum viable patch.  Update the existing TLI hook you added with the source parameter, let's get that landed, and then iterate.